usaspending_api/common/threaded_data_loader.py

```diff
@@ -0,0 +1,194 @@
+
+    # Loads data from a file using parameters set during creation of the loader
+    # The filepath parameter should be the string location of the file for use with open()
+    def load_from_file(self, filepath):
+        start_time = time.time()
+        print('Started processing file ' + filepath + ' at ' + str(start_time))
+        # Create the Queue object - this will hold all the rows in the CSV
+        row_queue = queue.Queue(maxsize=100)
+        self.exit_flag = False
+
+        # Spawn our threads
+        pool = []
+        while len(pool) < self.threads:
+            thread = DataLoaderThread("Thread-" + str(len(pool)), row_queue, self)
+            thread.start()
+            pool.append(thread)
+
+        with open(filepath) as csvfile:
+            reader = csv.DictReader(csvfile)
+            for row in reader:
+                row_queue.put(row)
+
+        while not row_queue.empty():
+            pass
+
+        self.exit_flag = True
+
+        for thread in pool:
+            thread.join()
+
+        if self.post_process_function is not None:
+            self.post_process_function()
+
+        end_time = time.time()
+        elapsed = end_time - start_time
+        print('Finished processing at ' + str(end_time) + '; Elapsed: ' + str(elapsed))
+
+
+class DataLoaderThread(threading.Thread):
+    def __init__(self, name, data_queue, loader):
+        threading.Thread.__init__(self)
+        self.name = name
+        self.data_queue = data_queue
+        self.loader = loader
+
+    def run(self):
+        print("Starting " + self.name)
+        self.process_data()
+        print("Exiting " + self.name)
+
+    def process_data(self):
+        # Make sure we weren't flagged to exit
+        while not self.loader.exit_flag:
+            if not self.data_queue.empty():
+                row = self.data_queue.get()
+                # Grab the collision field
+                update = False
+                collision_instance = None
+                model_instance = None
+                if self.loader.collision_field is not None:
+                    model_collision_field = self.loader.collision_field
+                    data_collision_field = model_collision_field
+                    # If it's a long label, we need to look it up to map to data
+                    if model_collision_field in settings.LONG_TO_TERSE_LABELS:
+                        data_collision_field = settings.LONG_TO_TERSE_LABELS[model_collision_field]
+                    query_object = Q({model_collision_field: row[data_collision_field]})
+                    collision_instance = self.loader.model_class.objects.filter(query_object).first()
+
+                    if collision_instance is not None:
+                        behavior = self.loader.collision_behavior
+                        if behavior is "delete":  # Delete the row from the data store and load new row
+                            collision_instnace.delete()
```

Comment:
Not yet - I am working on that. I have to make some tiny data sets to validate that values are actually updated.